core/server/primarykinds.go

```diff
@@ -18,6 +42,66 @@ func New() *PrimaryKinds {
 	return &kinds
 }
 
+// versionRank calculates a ranking score for a given version string. It takes into account
+// the prefix number (for example, in v1alpha2, 1 is the prefix) and the suffix
+// (for example, alpha or beta and their accompanying number, if present).
+// The function returns an integer representing the ranking score and any error that occurred.
+func versionRank(version string) (int, error) {
+	// special handling for the internal version
+	if version == "__internal" {
+		return 0, nil
+	}
+
+	// regex pattern to extract version information
+	re := regexp.MustCompile(`v(\d+)(alpha(\d+)|beta(\d+)|)$`)
+	match := re.FindStringSubmatch(version)
+	if match == nil {
+		return -1, fmt.Errorf("invalid version string: %s", version)
+	}
+
+	// get and process the version prefix and suffix
+	prefix, _ := strconv.Atoi(match[1])
+	suffixRank := 0
+	if match[3] != "" {
+		suffixRank, _ = strconv.Atoi(match[3])
+		suffixRank *= alphaWeight
+	} else if match[4] != "" {
+		suffixRank, _ = strconv.Atoi(match[4])
+		suffixRank *= betaWeight
+	} else {
+		suffixRank = stableWeight
+	}
+
+	// calculate and return the rank
+	return prefix*prefixWeight + suffixRank, nil
+}
+
+// compareGVK compares two GroupVersionKind (GVK) objects and determines
+// which has a higher version rank. The function returns an integer (-1, 0, or 1)
+// representing the comparison result, and any error that occurred.
+func compareGVK(gvk1, gvk2 schema.GroupVersionKind) (int, error) {
+	rank1, err1 := versionRank(gvk1.Version)
+	if err1 != nil {
+		return 0, fmt.Errorf("failed to parse gvk group: %s, version %s, kind:%s : %w", gvk1.Group, gvk1.Version, gvk1.Kind, err1)
+	}
+
+	rank2, err2 := versionRank(gvk2.Version)
+	if err2 != nil {
+		return 0, err2
+	}
+
+	if rank1 > rank2 {
+		return 1, nil
+	} else if rank1 < rank2 {
+		return -1, nil
+	} else {
+		return 0, nil
+	}
+}
+
+// DefaultPrimaryKinds generates a new PrimaryKinds object which contains
+// the highest version of each kind of known types from a Kubernetes scheme.
+// It returns the PrimaryKinds object and any error that occurred.
 func DefaultPrimaryKinds() (*PrimaryKinds, error) {
```

Comment:
do you think that adding a test case for this method would help to provide understanding and regression?